### DISCUSSION TITLE
Configuring renovate with pipenv and packagecloud

### DISCUSSION THREAD

**Discussion by 0inp** [renovatebot/renovate#17748] [Request Help]
Hi !

I'm trying to put renovate on a repo taht contains a python API.
The api is using pipenv. We've got two different Pipfile sources: pypi and packagecloud.
Here's what the pipfile looks like (a lot of deps have been removed for reading’s sake):
```
[[source]]
name = "ma"
url = "https://${PACKAGECLOUD_TOKEN}:@packagecloud.io/meilleursagents/ma-prod/pypi/simple"
verify_ssl = true

[[source]]
name = "pypi"
url = "https://pypi.org/simple"
verify_ssl = true

[dev-packages]
black = {version = "*", index = "pypi"}
factory-boy = {version = "*", index = "pypi"}
flake8 = {version = "*", index = "pypi"}


[packages]
alembic = {version = ">=1,<2", index = "pypi"}
ma-pubsub = {version = ">=2.5.0,<3", index = "ma"}


[requires]
python_version = "3.9"
```

My problem is that I can't make the specific "ma" deps lookup work. On the onboarding PR, I keep having this message:
```
⚠ Dependency Lookup Warnings ⚠

Please correct - or verify that you can safely ignore - these lookup failures before you merge this PR.

    Failed to look up dependency ma-pubsub

Files affected: Pipfile
```

 I add in my renovate config.js hostRules
 ```
     hostRules: [
       ...,
        {
            hostType: 'pipenv',
            matchHost: 'packagecloud.io/meilleursagents/ma-prod/pypi/simple',
            token: process.env.PACKAGECLOUD_TOKEN,
        },
    ],
```
Here I try putting the token in the `username` or `password` or `token` key. None of them work. Normally, I can access packagecloud through simple basic auth.
Moreover, we are running renovate as self-hosted bot through our CI pipeline with jenkins. So I pass the new token in environment variable in my jenkinsFile:
```
    environment {
        ...
        PACKAGECLOUD_TOKEN              = credentials('packagecloud-read-token')
    }
```

Finally, I keep having this log:
```
 INFO: Dependency extraction complete (repository=MeilleursAgents/B2CLeadRankingAPI)
       "baseBranch": "master",
       "stats": {
         "managers": {
           "dockerfile": {"fileCount": 1, "depCount": 1},
           "pipenv": {"fileCount": 1, "depCount": 43}
         },
         "total": {"fileCount": 2, "depCount": 44}
       }
DEBUG: Datasource unauthorized (repository=MeilleursAgents/B2CLeadRankingAPI)
       "datasource": "pypi",
       "packageName": "ma-pubsub",
       "url": "https://%24%7BPACKAGECLOUD_TOKEN%7D@packagecloud.io/meilleursagents/ma-prod/pypi/simple/ma-pubsub/"
DEBUG: Failed to look up dependency ma-pubsub (repository=MeilleursAgents/B2CLeadRankingAPI, packageFile=Pipfile, dependency=ma-pubsub)
```
It looks like the parsing/decode of the url is not correct, therefore, the bot can't replace the token with the env var, thus being able to access the deps.

I'm out if idea and any help will be welcomed.
Thanks in advance.

**Comment by rarkins** [comment]
`matchHost` should either be a host name like `packagecloud.io` or a fully qualified URL prefix like `https://packagecloud.io/meilleursagents/ma-prod/pypi/simple`. Please try to correct that first and retry

**Reply by 0inp** [reply]
I tried those solutions, but I keep having the same error

**Reply by rarkins** [reply]
Renovate won't do variable substitution for `${PACKAGECLOUD_TOKEN}` so suggest you remove that part and retest, along with the hostType fix.

**Reply by 0inp** [reply]
What is the `hostType` fix you're suggesting ? 
But there is something I don't understand. I'm not asking for renovate to do variable substitution. Normally it's for pipenv to do so. It's already working well inside my app when i'm doing calling `pipenv update` with the correct env vars set inside my container.
I just don't understand what is doing renovate that requires me to add the url of the `packagecloud.io` repositories. Everything needed is already set up in the `Pipfile`. All I need to do is to pass from jenkins (through the cronjob) to renovate the token env var required for `packagecloud`.

**Reply by rarkins** [reply]
Renovate doesn't call out to pip to perform the update. So it is not relevant whether pip/pioenv understands your variable substitution - it matters whether Renovate does

**Reply by 0inp** [reply]
Ok, so am I stuck ?
From what I understood, it is that :
- my environment variable inside the Pipfile is needed by pipenv in order to access my custom deps when I call `pipenv update`
- but renovate is not calling pipenv but it is still reading the Pipfile and is stuck with env var substitution 
is that correct ?

**Reply by rarkins** [reply]
You can configure a package rule to override registryUrls to a registry string without any variable embedded. That way Renovate can use its typical hostRules authentication for the URL
